File: scripts/keyword-check.pl
#! /usr/bin/env perl
# Copyright Gentoo Foundation 2007

use strict;
use warnings;
use Carp;

# create a hash of valid architectures
my $filename = 'profiles/arch.list';

open my $archlist_handle, '<', $filename or croak "Cannot open $filename : $!";

my %arch;
while ( <$archlist_handle> ) {
	chomp;
	$arch{$_} = ! m/^(?:\#|(?:prefix)?$)/xms;
}

close $archlist_handle or croak "couldn't close $archlist_handle";

# we have yet to print to the screen
my $first = 1;

# process ebuilds
while ( defined( my $ebuild = glob '*-*/*/*.ebuild' ) ) {
	@ARGV = $ebuild;
	while ( <> ) {
		if ( ?^KEYWORDS=?x ) {
			my ( @forbidden, @stable );

			# check keywords for validity
			foreach ( split /\s+/xms, ( split qr{"} )[1] ) {
				if ( ! s/^[-~]//xms ) {
					push @stable, $_;
				}
				if ( ! $arch{$_} ) {
					push @forbidden, $_;
				}
			}

			# print a report if necessary
			if ( @forbidden || @stable ) {
				if ( ! $first ) {
					print "\n";
				}
				else {
					$first=0;
				}
				$ebuild =~ s,/[^/]+/,/,xms;
				$ebuild = substr $ebuild, 0, length( $ebuild ) - 7;
				printf "EBUILD    : %s\n", $ebuild;
 				if ( @forbidden ) {
					printf "forbidden : %s\n", @forbidden;
				}
				if ( @stable ) {
					printf "stable    : %s\n", @stable;
				}
			}
		}
	} continue {
		if (eof) {
			reset;
		}
	}
}

# vim: set ts=4 sw=4 noexpandtab:

File: scripts/keyword-check.py
#! /usr/bin/env python
# Copyright Gentoo Foundation 2007

import glob
import fileinput
import re

archlist = []
first = 1

for arch in open( 'profiles/arch.list', 'r' ).readlines():
	arch = arch.rstrip()
	if arch != '' and arch != 'prefix' and not arch.startswith( '#' ):
		archlist.append( arch )

for file in glob.glob( '*/*/*.ebuild' ):
	for line in fileinput.input( [file] ):
		if line.startswith( 'KEYWORDS=' ):
			forbidden=[]
			stable=[]
			for kw in re.split( '\s+', line.split( '"' )[1] ):
				if kw == '': pass
				else:
					if not kw.startswith( '~' ) and not kw.startswith( '-' ):
						stable.append( kw )
					else:
						kw=kw.lstrip( '~-' )
					if not kw in archlist: forbidden.append( kw )
			if len( stable ) != 0 or len( forbidden ) != 0:
				if not first: print
				else: first=0
				print 'EBUILD    : %s' % re.sub( '/[^/]+/', '/', file )
				if len( forbidden ) != 0:
					print 'forbidden : %s' % ' '.join( forbidden )
				if len( stable ) != 0:
					print 'stable    : %s' % ' '.join( stable )

# vim: set ts=4 sw=4 noexpandtab:
